IndexMut incorrectly chosen instead of Index

``` rust
use std::ops::{Index, IndexMut};

struct S;

impl S { fn foo(&self) -> i32 { 0 } }

impl Index<usize> for S {
    type Output = S;
    fn index(&self, _: usize) -> &S { self }
}

impl IndexMut<usize> for S {
    fn index_mut(&mut self, _: usize) -> &mut S { self }
}

fn main() {
    let s = S;
    let _ = &mut s.foo(); // OK
    let _ = &mut (&s[0]).foo(); // OK
    let _ = &mut s[0].foo(); // error: cannot borrow immutable local variable `s` as mutable                                                                   
}
```

The whole example compiles it the `IndexMut` implementation is removed.

This [came up on Stack Overflow](https://stackoverflow.com/q/52989901/155423). A simplified example:

 ```rust
use std::cell::RefCell;

fn f(v: Vec<RefCell<u8>>) {
    let _t = &mut *v[0].borrow_mut();
}

fn main() {}
```

```
error[E0596]: cannot borrow immutable argument `v` as mutable
 --> src/main.rs:4:20
  |
3 | fn f(v: Vec<RefCell<u8>>) {
  |      - consider changing this to `mut v`
4 |     let _t = &mut *v[0].borrow_mut();
  |                    ^ cannot borrow mutably
```

Workarounds include:

1. Using extra braces:

    ```rust
    fn f(v: Vec<RefCell<u8>>) {
        let _t = &mut *{ v[0].borrow_mut() };
    }
    ```

1. Calling `Index` explicitly:

    fn f(v: Vec<RefCell<u8>>) {
        use std::ops::Index;
        let _t = &mut *v.index(0).borrow_mut();
    }

--- a/src/test/ui/inference/issue-28935.rs
+++ b/src/test/ui/inference/issue-28935.rs
@@ -1,0 +1,9 @@
+// check-pass
+
+use std::cell::RefCell;
+
+pub fn f(v: Vec<RefCell<u8>>) {
+    let _t = &mut *v[0].borrow_mut();
+}
+
+fn main() {}

